DataFormats/Detectors/CTP/src/ctpRateFetcher.cxx

```diff
@@ -0,0 +1,152 @@
+    return pileUpCorrection(triggerRate);
+  }
+  return -1;
+}
+double ctpRateFetcher::fetchCTPratesClassesNoPuCorr(uint64_t timeStamp, const std::string& className, int inputType)
+{
+  std::vector<ctp::CTPClass> ctpcls = mConfig->getCTPClasses();
+  std::vector<int> clslist = mConfig->getTriggerClassList();
+  int classIndex = -1;
+  for (size_t i = 0; i < clslist.size(); i++) {
+    if (ctpcls[i].name.find(className) != std::string::npos) {
+      classIndex = i;
+      break;
+    }
+  }
+  if (classIndex == -1) {
+    LOG(warn) << "Trigger class " << className << " not found in CTPConfiguration";
+    return -2.;
+  }
+  auto rate{mScalers->getRateGivenT(timeStamp * 1.e-3, classIndex, inputType)};
+  return rate.second;
+}
+double ctpRateFetcher::fetchCTPratesInputs(uint64_t timeStamp, int input)
+{
+  std::vector<ctp::CTPScalerRecordO2> recs = mScalers->getScalerRecordO2();
+  if (recs[0].scalersInps.size() == 48) {
+    return pileUpCorrection(mScalers->getRateGivenT(timeStamp * 1.e-3, input, 7).second);
+  } else {
+    LOG(error) << "Inputs not available";
+    return -1.;
+  }
+}
+double ctpRateFetcher::fetchCTPratesInputsNoPuCorr(uint64_t timeStamp, int input)
+{
+  std::vector<ctp::CTPScalerRecordO2> recs = mScalers->getScalerRecordO2();
+  if (recs[0].scalersInps.size() == 48) {
+    return mScalers->getRateGivenT(timeStamp * 1.e-3, input, 7).second;
+  } else {
+    LOG(error) << "Inputs not available";
+    return -1.;
+  }
+}
+
+double ctpRateFetcher::pileUpCorrection(double triggerRate)
+{
+  if (mLHCIFdata == nullptr) {
+    LOG(fatal) << "No filling" << std::endl;
+  }
+  auto bfilling = mLHCIFdata->getBunchFilling();
+  std::vector<int> bcs = bfilling.getFilledBCs();
+  double nbc = bcs.size();
+  double nTriggersPerFilledBC = triggerRate / nbc / constants::lhc::LHCRevFreq;
+  double mu = -std::log(1 - nTriggersPerFilledBC);
+  return mu * nbc * constants::lhc::LHCRevFreq;
+}
+
+void ctpRateFetcher::setupRun(int runNumber, o2::ccdb::BasicCCDBManager* ccdb, uint64_t timeStamp, bool initScalers)
+{
+  if (runNumber == mRunNumber) {
+    return;
+  }
+  mRunNumber = runNumber;
+  LOG(debug) << "Setting up CTP scalers for run " << mRunNumber;
+  std::map<string, string> metadata;
+  mLHCIFdata = ccdb->getSpecific<parameters::GRPLHCIFData>("GLO/Config/GRPLHCIF", timeStamp, metadata);
+  if (mLHCIFdata == nullptr) {
+    LOG(fatal) << "GRPLHCIFData not in database, timestamp:" << timeStamp;
+  }
+  metadata["runNumber"] = std::to_string(mRunNumber);
+  mConfig = ccdb->getSpecific<ctp::CTPConfiguration>("CTP/Config/Config", timeStamp, metadata);
+  if (mConfig == nullptr) {
+    LOG(fatal) << "CTPRunConfig not in database, timestamp:" << timeStamp;
+  }
+  if (initScalers) {
+    mScalers = ccdb->getSpecific<ctp::CTPRunScalers>("CTP/Calib/Scalers", timeStamp, metadata);
+    if (mScalers == nullptr) {
+      LOG(fatal) << "CTPRunScalers not in database, timestamp:" << timeStamp;
+    }
+    mScalers->convertRawToO2();
```

Comment:
@lietava if `mScalers == nullptr` this call is not protected, I would suggest this:
```
    if (mScalers) {
      mScalers->convertRawToO2();
    } else {
      LOG(fatal) << "CTPRunScalers not in database, timestamp:" << timeStamp;
    }
```